### PR TITLE
Trivia quiz

### PR DESCRIPTION
**Pull request by RohanJnr**
I did open a PR (See: #264, #266) about the same thing before but i closed it because i implemented a separate way of coding it.The code has drastically changed and i could even fix a few bugs which i couldn't in the old code.

Closes: #235

### CODE REVIEW

**Comment by sco1** [comment]
Do you intend to add more questions to the pool? There are only 5 in the retro category and none in the general category.

File: bot/seasons/evergreen/trivia_quiz.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    "Idk what rules to put in here haha."
+
+]
+
+
+class TriviaQuiz(commands.Cog):
+    """A cog for all quiz commands."""
+
+    def __init__(self, bot: commands.Bot) -> None:
+        print(type(bot))
```
[the remark this answers is not in the document]
Comment (RohanJnr) [reply bot/seasons/evergreen/trivia_quiz.py]: ops

File: bot/seasons/evergreen/trivia_quiz.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                if (
+                        ctx.author == self.game_owners[ctx.channel.id]
+                        or Roles.moderator in [role.id for role in ctx.author.roles]
+                ):
+                    await self.declare_winner(ctx.channel, player_data)
+                    self.game_status[ctx.channel.id] = False
+                else:
+                    await ctx.send(f"{ctx.author.mention}, you are not authorised to stop this game :ghost: !")
+        else:
+            self.game_status[ctx.channel.id] = False
```
[the remark this answers is not in the document]
Comment (RohanJnr) [reply bot/seasons/evergreen/trivia_quiz.py]: okay

**Comment by RohanJnr** [comment]
will add more questions soon

File: bot/seasons/evergreen/trivia_quiz.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    @staticmethod
+    def load_questions() -> dict:
+        """Load the questions from  json file."""
+        p = Path("bot", "resources", "evergreen", "trivia_quiz.json")
+        with p.open() as json_data:
+            questions = json.load(json_data)
+            return questions
+
+    @commands.command(name="quiz")
+    async def quiz_game(self, ctx: commands.Context, option: str, category: str = "general") -> None:
```
Comment (scragly) [inline bot/seasons/evergreen/trivia_quiz.py]: I would recommend removing the `option` argument. It's not necessary, as you can just ensure you check if there's a session already ongoing (as you do already in line 68).

File: bot/seasons/evergreen/trivia_quiz.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+                embed.title = f"Question #{len(done_question)}"
+                embed.description = q
+                await ctx.send(embed=embed)
+
+            def check(m: discord.Message) -> bool:
+                ratio = fuzz.ratio(answer.lower(), m.content.lower())
+                return ratio > 80 and m.channel == ctx.channel
+            try:
+                msg = await self.bot.wait_for('message', check=check, timeout=10)
+            except Exception as e:
```
Comment (scragly) [inline bot/seasons/evergreen/trivia_quiz.py]: This shouldn't be a general exception. Add the exceptions you're expecting to be raised here so you avoid suppressing unexpected errors.

File: bot/seasons/evergreen/trivia_quiz.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        embed = discord.Embed(colour=discord.Colour.blue())
+        embed.title = "Score Board"
+        embed.description = ""
+        for k, v in player_data.items():
+            embed.description += f"{k} : {v}\n"
+        await channel.send(embed=embed)
+
+    @staticmethod
+    async def declare_winner(channel: discord.TextChannel, player_data: dict) -> None:
+        """A function declare the winner of the quiz."""
```
Comment (scragly) [inline bot/seasons/evergreen/trivia_quiz.py]:
```suggestion
        """Announce the winner of the quiz in the game channel."""
```

Avoid prefixing "a function" to docstrings and try to use a more accurate description of what it's going to do, just like if someone asked you in person "what's this do if I call it?".

File: bot/seasons/evergreen/trivia_quiz.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+                winner = list(player_data.keys())[author_index]
+                winners_mention = winner.mention
+            await channel.send(
+                f"Congratz {winners_mention} :tada: "
+                f"You have won this quiz game with a grand total of {highest_points} points!!"
+            )
+
+    @property
+    def category_embed(self) -> discord.Embed:
+        """A function which returns an embed showing all avilable categories."""
```
Comment (scragly) [inline bot/seasons/evergreen/trivia_quiz.py]:
```suggestion
        """Build an embed showing all available trivia categories."""
```

File: bot/seasons/evergreen/trivia_quiz.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+        embed.title = "The available question categories are:"
+        embed.description = ""
+        for cat, description in self.categories.items():
+            embed.description += f"**- {cat.capitalize()}**\n{description.capitalize()}\n"
+        embed.set_footer(text="If not category is chosen, then a random one will be selected.")
+        return embed
+
+    @staticmethod
+    async def send_answer(channel: discord.TextChannel, question_dict: dict) -> None:
+        """A function that sends the answer."""
```
Comment (scragly) [inline bot/seasons/evergreen/trivia_quiz.py]:
```suggestion
        """Send the correct the answer of a question to the game channel."""
```

File: bot/seasons/evergreen/trivia_quiz.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+        if info != "":
+            embed.description += f"**Information**\n{info}\n"
+        embed.description += "Lets move to the next question."
+        await channel.send(embed=embed)
+
+
+def setup(bot: commands.Bot) -> None:
+    """Loading the cog."""
+    bot.add_cog(TriviaQuiz(bot))
+    logger.debug("TriviaQuiz cog loaded!")
```
Comment (scragly) [inline bot/seasons/evergreen/trivia_quiz.py]:
```suggestion
    logger.debug("TriviaQuiz cog loaded")
```
Keep consistent with other cases of similar code in the project.

File: bot/seasons/evergreen/trivia_quiz.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+    @staticmethod
+    def load_questions() -> dict:
+        """Load the questions from  json file."""
+        p = Path("bot", "resources", "evergreen", "trivia_quiz.json")
+        with p.open() as json_data:
+            questions = json.load(json_data)
+            return questions
+
+    @commands.command(name="quiz")
```
Comment (scragly) [inline bot/seasons/evergreen/trivia_quiz.py]: I'd suggest adding an alias of `trivia`.

File: bot/resources/evergreen/trivia_quiz.json [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+      "answer": "Mercury",
+      "info": "Mercury is the smallest planet in our solar system. It's just a little bigger than Earth's moon. It is the closest planet to the sun, but it's actually not the hottest. Venus is hotter."
+
+    },
+    {
+      "id": 128,
+      "question": "Is your country the smallest? Well it is if it is this perticular place, name the smallest country.",
+      "answer": "Vatican city",
+      "info": "With an area of 0.17 square miles (0.44 km2) and a population right around 1,000, Vatican City is the smallest country in the world, both in terms of size and population."
+
```
Comment (scragly) [inline bot/resources/evergreen/trivia_quiz.json]:
```suggestion
```
Random newline.

File: bot/resources/evergreen/trivia_quiz.json [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+      "answer": "Sherlock Holmes",
+      "info": "You guys know this very well, xD."
+
+    },
+    {
+      "id": 124,
+      "question": "When did the Second World War end?",
+      "answer": "1945",
+      "info": "World War 2 ended with the unconditional surrender of the Axis powers. On 8 May 1945, the Allies accepted Germany's surrender, about a week after Adolf Hitler had committed suicide. VE Day – Victory in Europe celebrates the end of the Second World War on 8 May 1945."
+
```
Comment (scragly) [inline bot/resources/evergreen/trivia_quiz.json]:
```suggestion
```
Random newline.

File: bot/resources/evergreen/trivia_quiz.json [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+      "answer": "elephant",
+      "info": "It is true that adult elephants can't jump. But there are other mammals that can't either, like sloths, hippos and rhinos. Although, unlike elephants, hippos and rhinos can have all four feet off the ground at the same time when they run."
+
+    },
+    {
+      "id": 123,
+      "question": "Who lived at 221B, Baker Street, London?",
+      "answer": "Sherlock Holmes",
+      "info": "You guys know this very well, xD."
+
```
Comment (scragly) [inline bot/resources/evergreen/trivia_quiz.json]:
```suggestion
```
Random newline.

File: bot/resources/evergreen/trivia_quiz.json [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+  "general":[
+    {
+      "id": 100,
+      "question": "Name the land of thousand lakes",
+      "answer": "Finland",
+      "info": "Finland is a country in Northern Europe. Sweden borders it to the northwest, Estonia to the south, Russia to the east, and Norway to the north. Finland is part of the European Union with its capital city being Helsinki. With a population of 5.5 million people, it has over 187,000 lakes. The thousands of lakes in Finland are the reason why the country's nickname is “the land of a thousand lakes."
+    },
+    {
+      "id": 101,
+      "question": "Winner of FIFA 2018",
```
[the remark this answers is not in the document]
Comment (RohanJnr) [reply bot/resources/evergreen/trivia_quiz.json]: Ahh I forgot the question mark

File: bot/resources/evergreen/trivia_quiz.json [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+      "id": 5,
+      "hints": ["The game was released in 1990.", "It was released on the SNES."],
+      "question": "What was the first game Yoshi appeared in?",
+      "answer": "Super Mario World"
+    }  
+  ],
+  "general":[
+    {
+      "id": 100,
+      "question": "Name the land of thousand lakes",
```
[the remark this answers is not in the document]
Comment (scragly) [reply bot/resources/evergreen/trivia_quiz.json]: Yes, you're right hem, thank you.

File: bot/resources/evergreen/trivia_quiz.json [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+    },
+    {
+      "id": 102,
+      "question": "Remind me the largest Ocean in the world.",
+      "answer": "Pacific",
+      "info": "The Pacific Ocean is the largest and deepest of the world ocean basins. Covering approximately 63 million square miles and containing more than half of the free water on Earth, the Pacific is by far the largest of the world's ocean basins."
+    },
+    {
+      "id": 103,
+      "question": "Have you heard of The Statue Of Liberty ? Who gifted it?",
```
[the remark this answers is not in the document]
Comment (RohanJnr) [reply bot/resources/evergreen/trivia_quiz.json]: I was trying to make the quiz a lil bit more interactive

File: bot/resources/evergreen/trivia_quiz.json [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+      "answer": "French",
+      "info": "The Statue of Liberty was a gift from the French people commemorating the alliance of France and the United States during the American Revolution. Yet, it represented much more to those individuals who proposed the gift. A photograph of Edouard de Laboulaye from the Galerie Contemporaine collection."
+    },
+    {
+      "id": 104,
+      "question": "Land Of The Rising Sun?",
```
[the remark this answers is not in the document]
Comment (RohanJnr) [reply bot/resources/evergreen/trivia_quiz.json]: "Which country", xD

File: bot/resources/evergreen/trivia_quiz.json [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+      "id": 102,
+      "question": "Remind me the largest Ocean in the world.",
+      "answer": "Pacific",
+      "info": "The Pacific Ocean is the largest and deepest of the world ocean basins. Covering approximately 63 million square miles and containing more than half of the free water on Earth, the Pacific is by far the largest of the world's ocean basins."
+    },
+    {
+      "id": 103,
+      "question": "Have you heard of The Statue Of Liberty ? Who gifted it?",
+      "answer": "French",
+      "info": "The Statue of Liberty was a gift from the French people commemorating the alliance of France and the United States during the American Revolution. Yet, it represented much more to those individuals who proposed the gift. A photograph of Edouard de Laboulaye from the Galerie Contemporaine collection."
```
[the remark this answers is not in the document]
Comment (RohanJnr) [reply bot/resources/evergreen/trivia_quiz.json]: Sorry

File: bot/resources/evergreen/trivia_quiz.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (scragly) [reply bot/resources/evergreen/trivia_quiz.json]: Fair, that works. I've edited the suggestion.

File: bot/resources/evergreen/trivia_quiz.json (hunk 17 above)
[the remark this answers is not in the document]
Comment (LordHemlock) [reply bot/resources/evergreen/trivia_quiz.json]: Also a minor note on the answer part of this question:  "French" wouldn't be a good answer.  Either "The French" or "France" would be a better way to put it.

[the remark this answers is not in the document]
Comment (scragly) [reply bot/resources/evergreen/trivia_quiz.json]: Yeah, good suggestion. `France` is appropriate.

File: bot/resources/evergreen/trivia_quiz.json [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+      "question": "Which is the only mammal that can’t jump?",
+      "answer": "elephant",
+      "info": "It is true that adult elephants can't jump. But there are other mammals that can't either, like sloths, hippos and rhinos. Although, unlike elephants, hippos and rhinos can have all four feet off the ground at the same time when they run."
+
+    },
+    {
+      "id": 123,
+      "question": "Who lived at 221B, Baker Street, London?",
+      "answer": "Sherlock Holmes",
+      "info": "You guys know this very well, xD."
```
[the remark this answers is not in the document]
Comment (RohanJnr) [reply bot/resources/evergreen/trivia_quiz.json]: Or maybe I can say "you guys should have known this very well xD"

[the remark this answers is not in the document]
Comment (scragly) [reply bot/resources/evergreen/trivia_quiz.json]: Nah, it's not consistent as no other question has similar banter in the info section. The rest are informative.

File: bot/seasons/evergreen/trivia_quiz.py (hunk 5 above)
Comment (RohanJnr) [reply bot/seasons/evergreen/trivia_quiz.py]: Yes, sir!

File: bot/resources/evergreen/trivia_quiz.json [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+    },
+    {
+      "id": 116,
+      "question": "the Vally Of The Kings is located in the country...?",
+      "answer": "Egypt",
+      "info": "The Valley of the Kings, also known as the Valley of the Gates of the Kings, is a valley in Egypt where, for a period of nearly 500 years from the 16th to 11th century BC, rock cut tombs were excavated for the pharaohs and powerful nobles of the New Kingdom (the Eighteenth to the Twentieth Dynasties of Ancient Egypt)."
+    },
+    {
+      "id": 117,
+      "question": "Diamonds are always nice in Minecraft, but did you know that there is a \"Diamond Capital in the World?\" can anyone name it?",
```
[the remark this answers is not in the document]
Comment (RohanJnr) [reply bot/resources/evergreen/trivia_quiz.json]: lol, you finally decided to keep an interactive start.

**Comment by RohanJnr** [comment]
I have implemented the given suggestions.

File: bot/resources/evergreen/trivia_quiz.json [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+    {
+      "id": 105,
+      "question": "What's known as the \"Playground of Europe\"?",
+      "answer": "Switzerland",
+      "info": "It comes from the title of a book written in 1870 by Leslie Stephen (father of Virginia Woolf) detailing his exploits of mountain climbing (not skiing) of which sport he was one of the pioneers and trekking or walking."
+    },
+    {
+      "id": 106,
+      "question": "Which country is known as the \"Land of Thunderbolt\"?",
+      "info": "\"Bhutan is known as the \\\"Land of Thunder Dragon\\\" or \\\"Land of Thunderbolt\\\" due to the violent and large thunderstorms that whip down through the valleys from the Himalayas. The dragon reference was due to people thinking the sparkling light of thunderbolts was the red fire of a dragon.\""
```
Comment (scragly) [inline bot/resources/evergreen/trivia_quiz.json]:
```suggestion
      "info": "Bhutan is known as the \"Land of Thunder Dragon\" or \"Land of Thunderbolt\" due to the violent and large thunderstorms that whip down through the valleys from the Himalayas. The dragon reference was due to people thinking the sparkling light of thunderbolts was the red fire of a dragon."
```

Appears to be an autoformatted mistake after pasting into pycharm.

File: bot/resources/evergreen/trivia_quiz.json [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+    },
+    {
+      "id": 121,
+      "question": "When did the First World War start?",
+      "answer": "1914",
+      "info": "The first world war began in August 1914. It was directly triggered by the assassination of the Austrian archduke, Franz Ferdinand and his wife, on 28th June 1914 by Bosnian revolutionary, Gavrilo Princip. This event was, however, simply the trigger that set off declarations of war."
+    },
+    {
+      "id": 122,
+      "question": "Which is the largest desert in the world (exclude cold deserts.) ?",
```
Comment (scragly) [inline bot/resources/evergreen/trivia_quiz.json]:
```suggestion
      "question": "Which is the largest hot desert in the world?",
```

File: bot/resources/evergreen/trivia_quiz.json [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+      "info": "It comes from the title of a book written in 1870 by Leslie Stephen (father of Virginia Woolf) detailing his exploits of mountain climbing (not skiing) of which sport he was one of the pioneers and trekking or walking."
+    },
+    {
+      "id": 106,
+      "question": "Which country is known as the \"Land of Thunderbolt\"?",
+      "info": "\"Bhutan is known as the \\\"Land of Thunder Dragon\\\" or \\\"Land of Thunderbolt\\\" due to the violent and large thunderstorms that whip down through the valleys from the Himalayas. The dragon reference was due to people thinking the sparkling light of thunderbolts was the red fire of a dragon.\""
+    },
+    {
+      "id": 107,
+      "question": "Which country is the largest producer of coffee?",
```
Comment (scragly) [inline bot/resources/evergreen/trivia_quiz.json]:
```suggestion
      "question": "Which country is the largest producer of tea in the world?",
```
Looks like you mixed tea and coffee questions up.

File: bot/resources/evergreen/trivia_quiz.json [hunk 23]
```diff
@@ -0,0 +1,221 @@
...
+      "answer": "China",
+      "info": "Tea is mainly grown in Asia, Africa, South America, and around the Black and Caspian Seas. The four biggest tea-producing countries today are China, India, Sri Lanka and Kenya. Together they represent 75% of world production."
+    },
+    {
+      "id": 108,
+      "question": "Which country is the largest producer of tea in the world?",
```
Comment (scragly) [inline bot/resources/evergreen/trivia_quiz.json]:
```suggestion
      "question": "Which country is the largest producer of coffee?",
```
Looks like you mixed tea and coffee questions up.

File: bot/resources/evergreen/trivia_quiz.json [hunk 24]
```diff
@@ -0,0 +1,221 @@
...
+    },
+    {
+      "id": 124,
+      "question": "When did the Second World War end?",
+      "answer": "1945",
+      "info": "World War 2 ended with the unconditional surrender of the Axis powers. On 8 May 1945, the Allies accepted Germany's surrender, about a week after Adolf Hitler had committed suicide. VE Day – Victory in Europe celebrates the end of the Second World War on 8 May 1945."
+    },
+    {
+      "id": 125,
+      "question": "What is the name of the largest Dam in the world?",
```
Comment (scragly) [inline bot/resources/evergreen/trivia_quiz.json]:
```suggestion
      "question": "What is the name of the largest dam in the world?",
```
Missed this one on the first review, whoops.

File: bot/resources/evergreen/trivia_quiz.json [hunk 25]
```diff
@@ -0,0 +1,221 @@
...
+    },
+    {
+      "id": 129,
+      "question": "What's the name of the largest bird?",
+      "answer": "Ostrich",
+      "info": "The largest living bird, a member of the Struthioniformes, is the ostrich (Struthio camelus), from the plains of Africa and Arabia. A large male ostrich can reach a height of 2.8 metres (9.2 feet) and weigh over 156 kilograms (344 pounds)."
+    },
+    {
+      "id": 130,
+      "question": "Expand GPRS",
```
Comment (scragly) [inline bot/resources/evergreen/trivia_quiz.json]:
```suggestion
      "question": "What does the acronym GPRS stand for?",
```
Another I seemed to have missed in my previous review.

File: bot/seasons/evergreen/trivia_quiz.py [hunk 26]
```diff
@@ -0,0 +1,234 @@
...
+
+    @staticmethod
+    def load_questions() -> dict:
+        """Load the questions from  json file."""
+        p = Path("bot", "resources", "evergreen", "trivia_quiz.json")
+        with p.open() as json_data:
+            questions = json.load(json_data)
+            return questions
+
+    @commands.command(name="quiz", aliases=["trivia", "tquiz"])
```
Comment (scragly) [inline bot/seasons/evergreen/trivia_quiz.py]:
```suggestion
    @commands.command(name="quiz", aliases=["trivia"])
```
I can't imagine anyone trying out `tquiz` on their first go unless they explicitly knew about it in advance, and even then, `quiz` is shorter and easier to remember.

File: bot/seasons/evergreen/trivia_quiz.py [hunk 27]
```diff
@@ -0,0 +1,234 @@
...
+        embed.title = "The available question categories are:"
+        embed.description = ""
+        for cat, description in self.categories.items():
+            embed.description += f"**- {cat.capitalize()}**\n{description.capitalize()}\n"
+        embed.set_footer(text="If not category is chosen, then a random one will be selected.")
+        return embed
+
+    @staticmethod
+    async def send_answer(channel: discord.TextChannel, question_dict: dict) -> None:
+        """Send the correct the answer of a question to the game channel."""
```
Comment (scragly) [inline bot/seasons/evergreen/trivia_quiz.py]:
```suggestion
        """Send the correct answer of a question to the game channel."""
```
My mistake, I was pretty worn out by the time I gave the suggestion for this docstring.

File: bot/seasons/evergreen/trivia_quiz.py [hunk 28]
```diff
@@ -0,0 +1,234 @@
...
+
+        if not self.game_status[ctx.channel.id]:
+            self.game_owners[ctx.channel.id] = ctx.author
+            self.game_status[ctx.channel.id] = True
+            start_embed = discord.Embed(colour=discord.Colour.red())
+            start_embed.title = "Quiz game Starting!!"
+            start_embed.description = "Each game consists of 5 questions.\n"
+            start_embed.description += "**Rules :**\nNo cheating and have fun!"
+            start_embed.set_footer(
+                text="Points for a question reduces by 25 after 10s or after a hint.Total time is 30s per question"
```
Comment (scragly) [inline bot/seasons/evergreen/trivia_quiz.py]:
```suggestion
                text="Points for a question reduces by 25 after 10s or after a hint. Total time is 30s per question"
```

**Comment by RohanJnr** [comment]
committed requested changes

**Pull request merged**